pkg/mgmt/backup/backup.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2020 The OpenEBS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"fmt"
+	"k8s.io/klog"
+	"time"
+
+	apis "github.com/openebs/zfs-localpv/pkg/apis/openebs.io/zfs/v1"
+	zfs "github.com/openebs/zfs-localpv/pkg/zfs"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+)
+
+// isDeletionCandidate checks if a zfs backup is a deletion candidate.
+func (c *BkpController) isDeletionCandidate(bkp *apis.ZFSBackup) bool {
+	return bkp.ObjectMeta.DeletionTimestamp != nil
+}
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two.
+func (c *BkpController) syncHandler(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the bkp resource with this namespace/name
+	bkp, err := c.bkpLister.ZFSBackups(namespace).Get(name)
+	if k8serror.IsNotFound(err) {
+		runtime.HandleError(fmt.Errorf("zfs backup '%s' has been deleted", key))
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	bkpCopy := bkp.DeepCopy()
+	err = c.syncBkp(bkpCopy)
+	return err
+}
+
+// enqueueBkp takes a ZFSBackup resource and converts it into a namespace/name
+// string which is then put onto the work queue. This method should *not* be
+// passed resources of any type other than ZFSBackup.
+func (c *BkpController) enqueueBkp(obj interface{}) {
+	var key string
+	var err error
+	if key, err = cache.MetaNamespaceKeyFunc(obj); err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	c.workqueue.Add(key)
+}
+
+// synBkp is the function which tries to converge to a desired state for the
+// ZFSBackup
+func (c *BkpController) syncBkp(bkp *apis.ZFSBackup) error {
+	var err error = nil
+	// ZFSBackup should be deleted. Check if deletion timestamp is set
+	if c.isDeletionCandidate(bkp) {
+		// reconcile for the Destroy error
+		err = zfs.DestoryBackup(bkp)
```

Comment:
Can this happen? looks like the finalizer gets added both for success/failed states. In case the failed state was due to backup not getting created, then destroy backup can fail too causing the backup  resource to be not deleted?

Reply:
no, at the destory time we do a check for the backup resource, so destory will not fail.